File: apps/api/src/scale_api/serial/protocol_parser.py
"""三种天平协议的字节流→WeightSample 解析。

线协议参考：
- mettler  : `S S  168.4521 g\\r\\n` / `S D  168.45 g\\r\\n`  (S=stable, D=dynamic)
- sartorius: `+ 168.4521 g \\r\\n` / `?  168.4521 g \\r\\n`     (?=unstable)
- generic  : 任何形如 `[+-]?d+(.d+)?\\s*(g|mg|kg)` 的行；按"读超时未变化"判 stable

每个解析器是 incremental state machine：feed(bytes) → list[WeightSample]，
内部缓冲未完整的行。
"""
from __future__ import annotations

import re
import time
from abc import ABC, abstractmethod
from typing import Literal

from .types import WeightSample

UnitT = Literal["g", "mg", "kg"]
_VALID_UNITS: set[str] = {"g", "mg", "kg"}


class ProtocolParser(ABC):
    """字节流增量解析器。"""

    def __init__(self, default_unit: UnitT = "g") -> None:
        self.buffer = bytearray()
        self.default_unit = default_unit

    def feed(self, chunk: bytes) -> list[WeightSample]:
        """喂字节，返回此次解析出的样本（可能为 0）。"""
        self.buffer.extend(chunk)
        out: list[WeightSample] = []
        while True:
            line, rest = self._extract_line(self.buffer)
            if line is None:
                break
            self.buffer = rest
            sample = self._parse_line(line)
            if sample is not None:
                out.append(sample)
        return out

    @staticmethod
    def _extract_line(buf: bytearray) -> tuple[bytes | None, bytearray]:
        """以 \\r\\n / \\n 为终止符切一行；不足则返回 (None, buf)。"""
        for term in (b"\r\n", b"\n", b"\r"):
            idx = buf.find(term)
            if idx != -1:
                return bytes(buf[:idx]), bytearray(buf[idx + len(term):])
        return None, buf

    @abstractmethod
    def _parse_line(self, line: bytes) -> WeightSample | None: ...


# Mettler protocol example (SICS):
#   "S S  168.4521 g"  -> stable
#   "S D  168.45 g"    -> dynamic
# Some firmwares prefix with extra 'S' (the SICS framing char).
_RE_METTLER = re.compile(
    rb"""^\s*S?\s*
        ([SD])\s+
        ([+-]?\d+(?:\.\d+)?)\s*
        (g|mg|kg)\s*$
    """,
    re.IGNORECASE | re.VERBOSE,
)


class MettlerParser(ProtocolParser):
    def _parse_line(self, line: bytes) -> WeightSample | None:
        line = line.strip()
        if not line:
            return None
        m = _RE_METTLER.match(line)
        if m is None:
            return None
        status, value, unit = m.group(1), m.group(2), m.group(3).decode().lower()
        if unit not in _VALID_UNITS:
            return None
        return WeightSample(
            value=float(value),
            unit=unit,  # type: ignore[arg-type]
            stable=status.upper() == b"S",
            raw=line.decode(errors="replace"),
            ts=time.time(),
        )


# Sartorius SBI protocol example:
#   "+ 168.4521 g"   -> stable
#   "? 168.4521 g"   -> unstable (? prefix)
#   "- 0.5 g"        -> stable, sign already in value
_RE_SARTORIUS = re.compile(
    rb"""^\s*([+\-?G])?\s*
        ([+-]?\d+(?:\.\d+)?)\s*
        (g|mg|kg)\s*$
    """,
    re.IGNORECASE | re.VERBOSE,
)


class SartoriusParser(ProtocolParser):
    def _parse_line(self, line: bytes) -> WeightSample | None:
        line = line.strip()
        if not line:
            return None
        m = _RE_SARTORIUS.match(line)
        if m is None:
            return None
        prefix, value, unit = m.group(1), m.group(2), m.group(3).decode().lower()
        if unit not in _VALID_UNITS:
            return None
        # `?` 前缀 = 不稳定；其它（包括无前缀）默认稳定
        stable = prefix not in (b"?",)
        return WeightSample(
            value=float(value),
            unit=unit,  # type: ignore[arg-type]
            stable=stable,
            raw=line.decode(errors="replace"),
            ts=time.time(),
        )


# Generic protocol: plain value + optional unit.
# Stability is determined by upstream timing (no change for read_timeout_ms).
# Examples:
#   "168.4521 g"
#   "ST,GS,+0123.456 g"   (CAS / OHAUS)
#   "100.5"               (no unit -> default)
_RE_GENERIC = re.compile(
    rb"""^[^\d+\-]*
        ([+-]?\d+(?:\.\d+)?)
        \s*
        (g|mg|kg)?
        .*$
    """,
    re.IGNORECASE | re.VERBOSE,
)


class GenericParser(ProtocolParser):
    def _parse_line(self, line: bytes) -> WeightSample | None:
        line = line.strip()
        if not line:
            return None
        m = _RE_GENERIC.match(line)
        if m is None:
            return None
        value = m.group(1)
        unit_match = m.group(2)
        unit = unit_match.decode().lower() if unit_match else self.default_unit
        if unit not in _VALID_UNITS:
            return None
        return WeightSample(
            value=float(value),
            unit=unit,  # type: ignore[arg-type]
            stable=True,  # 由 connection_manager 用时序判定
            raw=line.decode(errors="replace"),
            ts=time.time(),
        )


def make_parser(
    protocol_type: Literal["generic", "mettler", "sartorius"],
    default_unit: UnitT = "g",
) -> ProtocolParser:
    """工厂。"""
    if protocol_type == "mettler":
        return MettlerParser(default_unit=default_unit)
    if protocol_type == "sartorius":
        return SartoriusParser(default_unit=default_unit)
    return GenericParser(default_unit=default_unit)

File: apps/web/src/lib/serial/protocol-parser.ts
/**
 * 三种天平协议的字节流→WeightSample 解析（移植自 apps/api/src/scale_api/serial/protocol_parser.py）。
 *
 * 线协议参考：
 *  - mettler  : `S S  168.4521 g\r\n` / `S D  168.45 g\r\n`  (S=stable, D=dynamic)
 *  - sartorius: `+ 168.4521 g \r\n` / `?  168.4521 g \r\n`     (?=unstable)
 *  - generic  : 任何形如 `[+-]?d+(.d+)?\s*(g|mg|kg)` 的行；stable 由上层时序判定
 *
 * 每个解析器是 incremental state machine：feed(chunk) → 样本数组，未完整行存于 buffer。
 */

import type { WeightSample } from './adapter';

type Unit = WeightSample['unit'];
export type ProtocolType = 'generic' | 'mettler' | 'sartorius' | 'ohaus';

const VALID_UNITS = new Set<Unit>(['g', 'mg', 'kg']);
const TEXT_DECODER = new TextDecoder('latin1');

const LINE_TERMINATORS: ReadonlyArray<readonly number[]> = [
  [0x0d, 0x0a], // \r\n
  [0x0a], // \n
  [0x0d], // \r
];

function indexOfBytes(buf: Uint8Array, needle: readonly number[]): number {
  outer: for (let i = 0; i + needle.length <= buf.length; i++) {
    for (let j = 0; j < needle.length; j++) {
      if (buf[i + j] !== needle[j]) continue outer;
    }
    return i;
  }
  return -1;
}

function concatBytes(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

export abstract class ProtocolParser {
  protected buffer: Uint8Array = new Uint8Array(0);

  constructor(protected readonly defaultUnit: Unit = 'g') {}

  feed(chunk: Uint8Array): WeightSample[] {
    this.buffer = concatBytes(this.buffer, chunk);
    const out: WeightSample[] = [];
    while (true) {
      const ext = this.extractLine();
      if (!ext) break;
      const sample = this.parseLine(ext);
      if (sample) out.push(sample);
    }
    return out;
  }

  protected abstract parseLine(line: string): WeightSample | null;

  private extractLine(): string | null {
    let bestIdx = -1;
    let bestTermLen = 0;
    for (const term of LINE_TERMINATORS) {
      const idx = indexOfBytes(this.buffer, term);
      if (idx !== -1 && (bestIdx === -1 || idx < bestIdx)) {
        bestIdx = idx;
        bestTermLen = term.length;
      }
    }
    if (bestIdx === -1) return null;
    const lineBytes = this.buffer.slice(0, bestIdx);
    this.buffer = this.buffer.slice(bestIdx + bestTermLen);
    return TEXT_DECODER.decode(lineBytes);
  }

  protected makeSample(value: number, unit: Unit, stable: boolean, raw: string): WeightSample {
    return { value, unit, stable, raw, ts: Date.now() };
  }
}

const RE_METTLER = /^\s*S?\s*([SD])\s+([+-]?\d+(?:\.\d+)?)\s*(g|mg|kg)\s*$/i;

export class MettlerParser extends ProtocolParser {
  protected parseLine(line: string): WeightSample | null {
    const text = line.trim();
    if (!text) return null;
    const m = RE_METTLER.exec(text);
    if (!m) return null;
    const status = m[1] ?? '';
    const valueStr = m[2] ?? '';
    const unit = (m[3] ?? '').toLowerCase() as Unit;
    if (!VALID_UNITS.has(unit)) return null;
    return this.makeSample(Number(valueStr), unit, status.toUpperCase() === 'S', text);
  }
}

const RE_SARTORIUS = /^\s*([-+?G])?\s*([+-]?\d+(?:\.\d+)?)\s*(g|mg|kg)\s*$/i;

export class SartoriusParser extends ProtocolParser {
  protected parseLine(line: string): WeightSample | null {
    const text = line.trim();
    if (!text) return null;
    const m = RE_SARTORIUS.exec(text);
    if (!m) return null;
    const prefix = m[1];
    const valueStr = m[2] ?? '';
    const unit = (m[3] ?? '').toLowerCase() as Unit;
    if (!VALID_UNITS.has(unit)) return null;
    return this.makeSample(Number(valueStr), unit, prefix !== '?', text);
  }
}

const RE_GENERIC = /^[^\d+-]*([+-]?\d+(?:\.\d+)?)\s*(g|mg|kg)?.*$/i;

export class GenericParser extends ProtocolParser {
  protected parseLine(line: string): WeightSample | null {
    const text = line.trim();
    if (!text) return null;
    const m = RE_GENERIC.exec(text);
    if (!m) return null;
    const valueStr = m[1] ?? '';
    const unit = (m[2]?.toLowerCase() as Unit | undefined) ?? this.defaultUnit;
    if (!VALID_UNITS.has(unit)) return null;
    return this.makeSample(Number(valueStr), unit, true, text);
  }
}

export function makeParser(protocolType: ProtocolType, defaultUnit: Unit = 'g'): ProtocolParser {
  switch (protocolType) {
    case 'mettler':
      return new MettlerParser(defaultUnit);
    case 'sartorius':
      return new SartoriusParser(defaultUnit);
    case 'generic':
    case 'ohaus':
    default:
      return new GenericParser(defaultUnit);
  }
}
